devices/src/legacy/flash_memory.rs

```diff
@@ -0,0 +1,61 @@
+// Copyright 2022 Arm Limited (or its affiliates). All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::fmt;
+use std::result;
+use std::sync::Arc;
+use vm_memory::{
+    bitmap::AtomicBitmap, Address, GuestAddress, GuestMemoryRegion, GuestRegionMmap, MmapRegion,
+};
+type GuestMemoryMmap = vm_memory::GuestMemoryMmap<AtomicBitmap>;
+
+#[derive(Debug)]
+pub enum Error {
+    CreateFlashMemory(hypervisor::vm::HypervisorVmError),
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Error::CreateFlashMemory(e) => write!(f, "Failed to create flash memory: {}", e),
+        }
+    }
+}
+
+type Result<T> = result::Result<T, Error>;
+
+/// On AArch64, the UEFI binary requires a flash device at address 0.
+/// FlashMemory is a partially simulated flash. It uses memory for the storage,
+/// but doesn't implement the flash device spec.
+pub struct FlashMemory {
+    memory: GuestMemoryMmap,
+}
+
+impl FlashMemory {
+    pub fn new(
+        vm: &Arc<dyn hypervisor::Vm>,
+        addr: GuestAddress,
+        size: usize,
+        slot: u32,
+    ) -> Result<Self> {
+        let region = GuestRegionMmap::new(MmapRegion::new(size).unwrap(), addr).unwrap();
+        let mem_region = vm.make_user_memory_region(
+            slot,
+            region.start_addr().raw_value(),
+            region.len() as u64,
+            region.as_ptr() as u64,
+            false,
+            false,
+        );
+        vm.create_user_memory_region(mem_region)
+            .map_err(Error::CreateFlashMemory)?;
```

Comment:
@rbradford I don't know what's your opinion here, but it feels like we're crossing boundaries. This is a legacy device, meaning only the emulation of the device should be handled here (and since it doesn't emulate anything, there's not much that'd be needed here).
The `GuestMemory` part should be directly handled in the `vmm` crate, don't you think?

Reply:
I can be even simpler if we just place this piece of memory in `Vm` or `MemoryManager`. Do you mean that?

Reply:
Yes that's what I mean.

Reply:
Good, I will try.

Reply:
Done. It was placed in `DeviceManager` because anyhow it is more like a device. 
Although it is merely a block of memory, but it is different thing with what's handled in `MemoryManager` for the guest, that's why I didn't put it there.